### ISSUE TITLE
[FR]: Detect unused Python dependencies

### ISSUE THREAD

**Issue by jab** [aspect-build/rules_lint#296]
### What is the current behavior?

No linter is currently supported that detects unused Python dependencies.

### Describe the feature

Enable Python users to notice when they are taking dependencies that no targets are actually importing.

https://github.com/tweag/FawltyDeps is an existing tool that provides this for pip dependencies (UPDATE: and now also conda dependencies), as well as the opposite: flagging when code is importing a dependency that is not declared as a direct dependency.

Not sure if there is something more general-purpose (e.g. works with any kind of dependency, ideally even with any language, by leveraging Bazel).

**Comment by alexeagle** [comment]
Normally it's gazelle's job to manage deps by removing those which aren't imported (and have no #keep comment)
I'm not sure if it's desirable to have a second competing workflow for that...

**Comment by jab** [comment]
Does gazelle look in pyproject.toml and requirements.in files?

**Comment by arrdem** [comment]
Just to be precise here -- the ask is to have a tool which consumes a `requirements` source and lints that requirement source according to the criteria that any listed requirement must be ... directly? transitively? imported by the current source set? This could be a valid linter for a `requirements.in` list, but would be quite difficult to implement for a lockfile in part due to conditional importing/optional deps behavior.

**Comment by arrdem** [comment]
This would probably be a separate project unto itself with `rules_lint` integrating with it -- probably the wrong place for this discussion.

**Comment by jab** [comment]
Makes sense, will close this. If anyone comes across a separate project that does this and is able to link to it here, please do.

To answer your question: Direct dependencies, not transitive (so in Python this would [conventionally](https://github.com/theoremlp/rules_uv#pip_compile) be `requirements.in` files, not `requirements.txt` files). 

To clarify the idea: Build the directed graph where each source file is connected to the environment(s) it depends on, and each environment is connected to a declaration of that environment's direct dependencies. For each `import` in each source file, assert that the associated package appears in the direct dependency declaration of every environment reachable from that source file in the graph, or else fail with an "undeclared dependency" error. 

The other direction is worth checking too: For each package that appears in a dependency set, ensure there is a corresponding `import` that depends on that package in some source file that can be reached by traversing the graph, or else fail with an "unused dependency" error.

Ideally this could be implemented for multiple languages via the same interface, but getting this working just for (say) Python seems like a good starting point.
